Feature/post

leejaeeun59357 opened the pull request:
### 변경사항

**AS-IS**
- 회원가입, 로그인 기능

**TO-BE**
- 로그인이 되었을 때, 게시글 작성, 수정, 삭제 기능 추가
- 모든 사용자가 전체 게시글 조회, 키워드 입력하여 해당 게시글 검색 기능 추가
- 해당 게시글의 PK를 이용하여 삭제 및 수정
    - 해당 게시글 작성자만 삭제 수정 가능

### 테스트
- [x] 게시글 작성
- [x] 게시글 삭제
- [x] 전체 게시글 조회
- [x] 게시글 수정
- [x] 키워드 포함한 제목의 게시글 검색

ne2227 committed:
[fix] 예외 처리 추가

- 회원가입 입력 폼에 null로 입력되어도 저장되는 부분에 예외 발생시킴

ne2227 committed:
[fix] @AuthenticationPrincipal 어노테이션 사용 시 null 값만 들어오는 오류 수정

- 새로 객체를 생성하여 UserDetails 객체를 생성하지 않고 해당 사용자의 정보를 통해 UserDetails 객체 생성

ne2227 committed:
[feature] 게시글 작성 기능 구현

- @AuthenticaitonPrincipal 어노테이션을 사용하여 로그인한 사용자 아이디 사용
- 제목이나 내용이 null 값이면 에러 발생
- 이메일 인증이 완료되지 않은 작성자라면 에러 발생

ne2227 committed:
[feat] 에러 발생 시 해당 에러 메시지를 볼 수 있도록 기능 수정

ne2227 committed:
[refactor] 잘못된 에러 발생 이름 변경

ne2227 committed:
[refactor] 주석 추가

ne2227 committed:
[feat] 게시물 삭제 기능 구현

- 로그인한 사용자의 ID와 게시물 제목으로 해당하는 게시물을 찾은 후 게시물 중 첫번째 게시물 삭제하는 기능 구현
- 만약 해당하는 게시물 없다면 에러 발생

ne2227 committed:
[refactor] class 이름 변경

ne2227 committed:
[refactor] 주석처리되어있던 사용하지 않는 기능 삭제

Commit:
[chore] 게시물 저장, 삭제 시 @Transactional 어노테이션 추가

Commit:
[feat] 전체 게시물 기능 구현

- 전체 게시물을 최신순으로 10개씩 조회할 수 있는 기능 구현
- 게시물 조회 시 인증 필요없이 조회할 수 있도록 함

Commit:
[refactor] 주석 추가

Commit:
[feat] 게시물 수정 기능 구현

- 제목과 작성자를 통해 해당 게시물을 수정할 수 있는 기능 구현
- 변경 전 제목, 변경 후 제목, 내용을 입력 받을 때 null 값이 들어오면 에러 발생

Commit:
[feat] 게시글 제목 검색 기능 구현

- 키워드를 입력받아 해당 키워드가 있는 게시글 제목 조회
- 해당 게시글이 10개 이상일 때 작성한 순으로 10개씩 보여지도록 함
- 키워드가 null 일 때 에러 발생
- 게시글 검색 기능은 인증 없이도 접근 가능하도록 함

Commit:
Update README.md

- 기능 구현 완료한 부분 체크 표시
- 기능 구현 시 변경된 부분 반영
- 해당 기능 구현 못했을 때 추가 기능 구현 예정으로 작성

Commit:
[refactor] 주석 추가

Commit:
[test] 게시글 조회 테스트

- 전체 게시글 조회, 게시글 제목 검색 테스트 scratch파일 작성

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/application/ModifyApplication.java:
```diff
@@ -0,0 +1,42 @@
+
+  public PostDto modifyPost(ModifyForm modifyForm, String writerId) {
+    // 변경 전 제목 null 일 때,
+    if (modifyForm.getBeforeSubject() == null || modifyForm.getBeforeSubject().isEmpty()) {
+      throw new PostCustomException(PostErrorCode.SUBJECT_IS_EMPTY);
+    }
```
변경전 제목이 null 인 경우를 체크할 필요가 있을까요? 변경후 제목이 null인지만 체크를 한다면, 변경전 제목이 null로 들어갈 일이 없지 않을까 싶어요

leejaeeun59357 replied:
넵! 수정하겠습니다!

leejaeeun59357 replied:
해당 게시글 수정 시
- 게시글 제목으로 수정 -> 게시글 ID로 게시글 찾아 수정

으로 변경하여 이 부분 삭제하였습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/application/ModifyApplication.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    // 변경 후 제목 null 일 때,
+    if (modifyForm.getAfterSubject() == null || modifyForm.getAfterSubject().isEmpty()) {
```
이 부분을 `ObjectUtils.isEmpty`를 쓰면 null safe하게 체크 가능할 것 같습니다

leejaeeun59357 replied:
넵! 수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/application/ModifyApplication.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    // 해당 제목의 게시물 찾기
+    Post post = postService.findPosts(writerId, modifyForm.getBeforeSubject());
```
수정하고자 하는 글을 제목으로 찾는다면, 같은 제목의 글을 여러번 쓰는 경우는 제대로된 게시물을 찾을수 없을 것 같습니다. 
이 경우는 수정하고자 하는 글의 PK 값을 API 사용처에서 요청에 포함해야할 것 같네요

leejaeeun59357 replied:
게시글 수정 시 
- 게시글 제목 입력 -> 게시글 PK값 사용

으로 변경하였습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/application/RemoveApplication.java:
```diff
@@ -0,0 +1,22 @@
+
+  @Transactional
+  public String removePost(String writerId, String subject) {
+    Post post = postService.findPosts(writerId, subject);
+    postService.removePost(post);
```
굳이 post를 find할 필요 없이, post pk 값을 받아서 바로 지우면 좋을 것 같습니다. 
select하여 나온 post entity로 어떤 연산을 수행하고 있지는 않으니 select가 아예 발생하지 않도록 native query로 작성해주시면 더욱 효율적인 코드가 될듯 합니다

leejaeeun59357 replied:
postID를 통해 조회한 post와 로그인한 사용자가 동일인물인지 확인하는 과정이 필요하다고 생각하여 post를 find 하였습니다!
이후 동일하다면 삭제하는 과정을 진행하였습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/service/PostService.java:
```diff
@@ -0,0 +1,125 @@
+   * @param writerId
+   * @return
+   */
+  public boolean isEmailVerified(String writerId) {
+    User user = findUser(writerId);
+    return user.getVerifiedStatus();
+  }
```
이 부분은 postService보단 userService에 좀더 어울리는 메소드인것 같네요. post와는 도메인적으로 연관이 없어보입니다.

leejaeeun59357 replied:
넵! 수정했습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/service/PostService.java:
```diff
@@ -0,0 +1,125 @@
+   * @param subject  게시물 제목
+   * @return
+   */
+  public Post findPosts(String writerId, String subject) {
```
- post 1개만 반환을 한다면 findPosts가 아닌 findPost가 맞는 메소드명일것 같습니다
- 제목으로 검색을 하는 것이 아닌, 제목을 통해 특정 게시물을 반환해야하는 목적이라면 (반환받고자 하는 게시물이 명확히 존재하는 거라면) PK 값으로 select하는 것이 좋을 것 같습니다. 위에서 언급한것처럼 같은 제목으로 글을 여러개를 쓴다면 해당 제목으로 이 메소드를 수행했을때 원하는 게시글을 return 받지 못할 수 있을 것 같습니다

leejaeeun59357 replied:
예! 그래서 게시글을 찾을 때 해당 게시글의 PK를 이용하여 찾는 것으로 변경하였습니다.
또한 1개의 Post를 반환하기 때문에 메소드명 변경 하였습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/service/PostService.java:
```diff
@@ -0,0 +1,125 @@
+   * @param writerId
+   * @return
+   */
+  public User findUser(String writerId) {
+    User user = userRepository.findByLoginId(writerId)
+        .orElseThrow(() -> new PostCustomException(PostErrorCode.NOT_FOUND_USER));
+
+    return user;
```
이 메소드도 userService에 좀더 어울리는 메소드로 보입니다

leejaeeun59357 replied:
넵! 수정했습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Post/service/PostService.java:
```diff
@@ -0,0 +1,125 @@
+  public List<PostDto> searchPost(String keyword, Pageable pageable) {
+    Page<Post> postLists = postRepository.findBySubjectContaining(keyword, pageable);
+
+    List<PostDto> postDtos = new ArrayList<>();
+    for (Post post : postLists) {
+      postDtos.add(PostDto.entityToDto(post));
+    }
```
list를 만들어서 여기에 add를 하는 방식보다, 요즘에는 선언과 할당을 같이하여 변수를 final화하여 사용하는 것을 권장하고 있습니다. 위 경우도 아래와 같이 코드를 수정해볼수 있을 것 같네요. 불필요한 선언을 줄이고 가독성을 높일 수 있을 것 같습니다
~~~java
return postRepository.findBySubjectContaining(keyword, pageable).stream().map(it -> PostDto.entityToDto(it)).collect(Collectors.toList());
~~~

leejaeeun59357 replied:
수정했습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/domain/model/User.java:
```diff
@@ -37,7 +37,7 @@ public class User extends BaseEntity {
   private String email;
 
-  @Enumerated(EnumType.STRING)
+//  @Enumerated(EnumType.STRING)
```
주석처리 된 이유가 있나요?

leejaeeun59357 replied:
아니요! 제가 주석처리 해놓고 해제하는 것을 깜빡했던 것 같습니다! 주석 해제 처리했습니다!!

mingming-mentor commented on src/main/java/org/board/exercise_board/User/application/SignUpApplication.java:
```diff
@@ -22,6 +22,16 @@ public class SignUpApplication {
   private final EmailService emailService;
 
   public UserDto signup(SignUpForm signUpForm) {
+    if (signUpForm.getEmail() == null) {
+      throw new CustomException(ErrorCode.EMAIL_IS_NULL);
+    }
+    if (signUpForm.getLoginId() == null) {
+      throw new CustomException(ErrorCode.LOGIN_ID_IS_NULL);
+    }
+    if (signUpForm.getPassword() == null) {
+      throw new CustomException(ErrorCode.PASSWORD_IS_NULL);
+    }
```
empty 체크도 같이 하면 좋을 것 같구요, validation annotation을 이용해도 좋을 것 같습니다. 
@NotBlank 같은 anntoation이요
https://sanghye.tistory.com/36

leejaeeun59357 replied:
넵! 수정했습니다!

mingming-mentor reviewed (approved):
리뷰드렸습니다. 수고 많으셨어요

Commit:
[refactor] 코드 리팩토링

- empty 찾는 조건 리팩토링
- 게시글 서비스 중 사용자 서비스와 관련된 부분은 UserService 로 이동

Commit:
[feat] 회원가입 시 입력값 검사 조건 추가

Commit:
[feat] 회원가입 입력폼에 조건 추가

Commit:
[refactor] @Valid 어노테이션 추가

Commit:
[feat] 게시글 삭제, 수정 시 Post의 PK값 이용으로 변경

- 해당 게시글을 찾을 때 Post의 PK값으로 찾기로 변경
- 게시글 수정 시, 변경 전 제목 입력 삭제
- 게시글 수정 입력 폼에 이전 제목 입력 -> PostID값 입력
- 게시글 수정, 삭제 시 로그인한 사용자와 작성자가 동일 사용자인지 확인 후 다르다면 에러 발생

Commit:
[refactor] 게시글 삭제 관련 메서드 이름 중remove -> delete로 메서드명 통일

Commit:
[refactor] 주석처리 되었던 @Enumerated 주석 해제

Commit:
[refactor] 게시글 ID를 RequestParam으로 요청받기로 변경

seowonn commented on src/main/java/org/board/exercise_board/User/application/SignUpApplication.java:
```diff
@@ -22,6 +22,16 @@ public class SignUpApplication {
   private final EmailService emailService;
 
   public UserDto signup(SignUpForm signUpForm) {
+    if (signUpForm.getEmail() == null) {
```
'spring-boot-starter-validation' 에서 @NotBlank 는 null 과 "" 과 " " 모두 허용하지 않는다고 합니다!

leejaeeun59357 replied:
좋은 정보 감사합니다!!!

Commit:
[feat] 회원가입 시 id,pw,email 유효성 검사 에러 발생 메시지 커스터마이징
